File: 卡码/代码随想录/1/06-二叉树/Js/01-二叉树遍历方式/LeetCode_0117.js
//117. 填充每个节点的下一个右侧节点指针 II
function Node(val, left, right, next) {
  this.val = val === undefined ? null : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
  this.next = next === undefined ? null : next;
};

const connect = function(root) {
  const q=new Array();
  let pre,cur;
  if(root) q.push(root);
  while(q.length>0){
    let size=q.length;
    for(let i=0;i<size;i++){
      cur=q.shift();
      if(i!=0) pre.next=cur;
      pre=cur;
      if(cur.left) q.push(cur.left);
      if(cur.right) q.push(cur.right);
    }
  }
  return root;
};

File: 卡码/代码随想录/1/06-二叉树/Go/01-二叉树遍历方式/LeetCode_0117.go
package main

import "container/list"

type Node struct {
	Val int
	Left *Node
  Right *Node
  Next *Node
}

func connect(root *Node) *Node {
	q:=list.New()
	var cur,pre *Node
	if root!=nil {
		q.PushBack(root)
	}
	for q.Len()>0 {
		size:=q.Len()
		for i:=0;i<size;i++ {
			if i==0 {
				pre=q.Remove(q.Front()).(*Node)
				cur=pre
			}else{
				cur=q.Remove(q.Front()).(*Node)
				pre.Next=cur
				pre=cur
			}
			if cur.Left!=nil {
				q.PushBack(cur.Left)
			}
			if cur.Right!=nil {
				q.PushBack(cur.Right)
			}
		}
		pre.Next=nil
	}
	return root
}
